spark/src/main/java/org/apache/iceberg/spark/source/SparkWriterFactory.java

```diff
@@ -0,0 +1,240 @@
+      Schema positionDeleteSchema = DeleteSchemaUtil.posDeleteSchema(positionDeleteRowSchema());
+      this.positionDeleteSparkType = SparkSchemaUtil.convert(positionDeleteSchema);
+    }
+
+    return positionDeleteSparkType;
+  }
+
+  static class Builder {
+    private final Table table;
+    private FileFormat dataFileFormat;
+    private Schema dataSchema;
+    private StructType dataSparkType;
+    private SortOrder dataSortOrder;
+    private FileFormat deleteFileFormat;
+    private int[] equalityFieldIds;
+    private Schema equalityDeleteRowSchema;
+    private StructType equalityDeleteSparkType;
+    private SortOrder equalityDeleteSortOrder;
+    private Schema positionDeleteRowSchema;
+    private StructType positionDeleteSparkType;
+    private SortOrder positionDeleteSortOrder;
+
+    Builder(Table table) {
+      this.table = table;
+
+      Map<String, String> properties = table.properties();
+
+      String dataFileFormatName = properties.getOrDefault(DEFAULT_FILE_FORMAT, DEFAULT_FILE_FORMAT_DEFAULT);
+      this.dataFileFormat = FileFormat.valueOf(dataFileFormatName.toUpperCase(Locale.ENGLISH));
+
+      String deleteFileFormatName = properties.getOrDefault(DELETE_DEFAULT_FILE_FORMAT, dataFileFormatName);
+      this.deleteFileFormat = FileFormat.valueOf(deleteFileFormatName.toUpperCase(Locale.ENGLISH));
+    }
+
+    Builder dataFileFormat(FileFormat newDataFileFormat) {
+      this.dataFileFormat = newDataFileFormat;
+      return this;
+    }
+
+    Builder dataSchema(Schema newDataSchema) {
+      this.dataSchema = newDataSchema;
+      return this;
+    }
+
+    Builder dataSparkType(StructType newDataSparkType) {
+      this.dataSparkType = newDataSparkType;
+      return this;
+    }
+
+    Builder dataSortOrder(SortOrder newDataSortOrder) {
+      this.dataSortOrder = newDataSortOrder;
+      return this;
+    }
+
+    Builder deleteFileFormat(FileFormat newDeleteFileFormat) {
+      this.deleteFileFormat = newDeleteFileFormat;
+      return this;
+    }
+
+    Builder equalityFieldIds(int[] newEqualityFieldIds) {
+      this.equalityFieldIds = newEqualityFieldIds;
+      return this;
+    }
+
+    Builder equalityDeleteRowSchema(Schema newEqualityDeleteRowSchema) {
+      this.equalityDeleteRowSchema = newEqualityDeleteRowSchema;
+      return this;
+    }
+
+    Builder equalityDeleteSparkType(StructType newEqualityDeleteSparkType) {
+      this.equalityDeleteSparkType = newEqualityDeleteSparkType;
+      return this;
+    }
+
+    Builder equalityDeleteSortOrder(SortOrder newEqualityDeleteSortOrder) {
+      this.equalityDeleteSortOrder = newEqualityDeleteSortOrder;
+      return this;
+    }
+
+    Builder positionDeleteRowSchema(Schema newPositionDeleteRowSchema) {
+      this.positionDeleteRowSchema = newPositionDeleteRowSchema;
+      return this;
+    }
+
+    Builder positionDeleteSparkType(StructType newPositionDeleteSparkType) {
```

Comment:
I'd like to emphasize that this is a complete type (includes path and pos), not just the row schema.
The assumption here is that we will validate the write schema before constructing a writer factory. Hence, we will already have the Spark type and may skip one more round of conversion from Iceberg to Spark.

Reply:
So for Hive we suppose to add our own method?

Reply:
Oh.. It is only for SparkWriterFactory..

Reply:
Yeah, correct. One of the reasons I did not introduce a common builder is because we still need some engine specific methods. Let me know if there are any ideas.

Reply:
I doubt people will set those parameters: 
* dataSparkType
* equalityDeleteSparkType
* positionDeleteSparkType

Those spark types are always transformed from iceberg schema, such as `dataSchema`, `equalityDeleteRowSchema`, `positionDeleteRowSchema`.  There's no much meaning that people convert those iceberg schema to a spark type, and then set up those spark types as the factory's default implementation. 

Especially, the `positionDeleteSparkType` will need to include all the file_path, row_offset and positional delete row schema.  It is actually a relatively high threshold for users.  

I'd recommend to remove them in this PR and add them back if someone think we have to provide those methods for some purposes.

Reply:
@openinx, all Spark types are actually optional and are not required. In fact, we don't set them in tests and they are derived from Iceberg values by `SparkWriterFactory`.

```
private StructType dataSparkType() {
  if (dataSparkType == null) {
    Preconditions.checkNotNull(dataSchema(), "Data schema must not be null");
    this.dataSparkType = SparkSchemaUtil.convert(dataSchema());
  }

  return dataSparkType;
}
```

What I am trying to say that we usually perform the write validation in Spark and we already have the Spark type.

```
Schema writeSchema = SparkSchemaUtil.convert(table.schema(), dsSchema);
TypeUtil.validateWriteSchema(table.schema(), writeSchema,
    checkNullability(spark, options), checkOrdering(spark, options));
```

Specifically, the existing `SparkAppenderFactory` actually requires the data Spark type right now.

```
SparkAppenderFactory appenderFactory = SparkAppenderFactory.builderFor(table, writeSchema, dsSchema).build();
```

In this PR, I made the Spark types optional. I think we will set them in general as we will always perform the write validation and will have correct Spark types.

Reply:
Okay,  make sense !